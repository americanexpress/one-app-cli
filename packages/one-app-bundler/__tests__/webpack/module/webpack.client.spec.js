/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

/* eslint-disable global-require */
const { validateWebpackConfig } = require('../../../test-utils');
const getConfigOptions = require('../../../utils/getConfigOptions');
const configGenerator = require('../../../webpack/module/webpack.client');

jest.mock('../../../utils/getConfigOptions', () => jest.fn(() => ({ purgecss: {} })));

jest.spyOn(process, 'cwd').mockImplementation(() => __dirname.split('/__tests__')[0]);

describe('webpack/module.client', () => {
  let originalNodeEnv;
  beforeAll(() => { originalNodeEnv = process.env.NODE_ENV; });
  afterAll(() => { process.env.NODE_ENV = originalNodeEnv; });

  it('should export valid webpack config', () => {
    const webpackConfig = configGenerator();
    return validateWebpackConfig(webpackConfig);
  });

  it('should provide the envName to babel', () => {
    const modernWebpackConfig = configGenerator('modern');
    const legacyWebpackConfig = configGenerator('legacy');
    expect(modernWebpackConfig.module.rules[2].use[0].options.envName).toBe('modern');
    expect(legacyWebpackConfig.module.rules[2].use[0].options.envName).toBe('legacy');
  });

  it('should warn for perf budget violations in development', () => {
    process.env.NODE_ENV = 'development';
    const webpackConfig = configGenerator();
    expect(webpackConfig.performance).toMatchObject({
      maxAssetSize: 250e3,
      hints: 'warning',
    });
  });

  it('should error for perf budget violations in production', () => {
    process.env.NODE_ENV = 'production';
    const webpackConfig = configGenerator();
    expect(webpackConfig.performance).toMatchObject({
      maxAssetSize: 250e3,
      hints: 'error',
    });
  });

  it('should accept a custom perf budget', () => {
    process.env.NODE_ENV = 'production';
    getConfigOptions.mockReturnValueOnce({ performanceBudget: 4103, purgecss: {} });
    const webpackConfig = configGenerator();
    expect(webpackConfig.performance).toMatchObject({
      maxAssetSize: 4103,
      maxEntrypointSize: 4103,
      hints: 'error',
    });
  });

  it('should define global.BROWSER to be true', () => {
    const webpackConfig = configGenerator();
    expect(webpackConfig).toHaveProperty('plugins', expect.any(Array));
    expect(webpackConfig.plugins).toContainEqual({ definitions: { 'global.BROWSER': 'true' } });
  });
});
