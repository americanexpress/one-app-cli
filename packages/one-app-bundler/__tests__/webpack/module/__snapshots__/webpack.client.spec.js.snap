// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack/module.client should generate full list of fall backs 1`] = `
Object {
  "assert": "/mock/path/for/request/assert",
  "buffer": "/mock/path/for/request/buffer",
  "console": "/mock/path/for/request/console-browserify",
  "constants": "/mock/path/for/request/constants-browserify",
  "crypto": "/mock/path/for/request/crypto-browserify",
  "domain": "/mock/path/for/request/domain-browser",
  "events": "/mock/path/for/request/events",
  "http": "/mock/path/for/request/stream-http",
  "https": "/mock/path/for/request/https-browserify",
  "os": "/mock/path/for/request/os-browserify/browser",
  "path": "/mock/path/for/request/path-browserify",
  "process": "/mock/path/for/request/process/browser",
  "punycode": "/mock/path/for/request/punycode",
  "querystring": "/mock/path/for/request/querystring-es3",
  "stream": "/mock/path/for/request/stream-browserify",
  "string_decoder": "/mock/path/for/request/string_decoder",
  "sys": "/mock/path/for/request/util",
  "timers": "/mock/path/for/request/timers-browserify",
  "tty": "/mock/path/for/request/tty-browserify",
  "url": "/mock/path/for/request/url",
  "util": "/mock/path/for/request/util",
  "vm": "/mock/path/for/request/vm-browserify",
  "zlib": "/mock/path/for/request/browserify-zlib",
}
`;
